test/functional/users_controller_test.rb

```diff
@@ -364,6 +365,42 @@ class UsersControllerTest < ActionController::TestCase
     assert_response :unprocessable_entity
   end
 
+  context "when user is logged in" do
+    setup do
+      @previous_url = "/bookmarks"
+      get :login, set_session_user
```

Comment:
I think this can be removed, it doesn't appear to be testing anything. The main requests are in the body of the test.